runtime/vm/segment.c

```diff
@@ -880,3 +890,52 @@ printSegments(J9MemorySegment *s, void* data)
 	printf("\n");
 }
 #endif /* DEBUG_PRINT_SEGMENT_TYPES */
+
+
+BOOLEAN
+disclaimClassMemory(J9JavaVM *vm, UDATA flags)
+{
+	BOOLEAN rc = TRUE;
+#if defined(LINUX)
+	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+	J9ClassLoaderWalkState walkState = {0};
+	J9ClassLoader *classLoader = vmFuncs->allClassLoadersStartDo(&walkState, vm, 0);
+	BOOLEAN romClass = J9_ARE_ANY_BITS_SET(flags, J9VM_DISCLAIMCLASSMEMORY_ROM);
+	BOOLEAN constRAMClass = J9_ARE_ANY_BITS_SET(flags, J9VM_DISCLAIMCLASSMEMORY_RAM);
+	BOOLEAN allRAMClass = J9_ARE_ANY_BITS_SET(flags, J9VM_DISCLAIMCLASSMEMORY_ALL_RAM);
+	BOOLEAN allClassMemory = romClass && constRAMClass && allRAMClass;
+	BOOLEAN stopOnFirstFailure = J9_ARE_ANY_BITS_SET(flags, J9VM_DISCLAIMCLASSMEMORY_STOP_ON_FIRST_FAILURE);
+
+	while (NULL != classLoader) {
+		J9MemorySegment *segment = classLoader->classSegments;
+		while (NULL != segment) {
+			BOOLEAN isROM = J9_ARE_ANY_BITS_SET(segment->type, MEMORY_TYPE_ROM_CLASS);
+			BOOLEAN isRAM = J9_ARE_ANY_BITS_SET(segment->type, MEMORY_TYPE_RAM_CLASS);
+			if (allClassMemory
+			|| (allRAMClass && isRAM)
+			|| (constRAMClass && isRAM && J9_ARE_NO_BITS_SET(segment->type, MEMORY_TYPE_RAM_CLASS_SUB4G))
+			|| (romClass && isROM)
+			) {
+				I_32 result = madvise(segment->heapBase, segment->size, MADV_PAGEOUT);
+				Trc_Segment_disclaimClassMemory_result(segment, segment->heapBase,  segment->size, segment->type, segment->baseAddress, result);
+				if (-1 == result) {
+					rc = FALSE;
+					if (stopOnFirstFailure) {
+						goto done;
+					}
+				}
+			}
+			segment = segment->nextSegmentInClassLoader;
+		}
+		classLoader = vmFuncs->allClassLoadersNextDo(&walkState);vmFuncs->allClassLoadersNextDo(&walkState);
+	}
+
+done:
+	vmFuncs->allClassLoadersEndDo(&walkState);
+#else /* defined(LINUX) */
+	/* Support for other platforms may be added in the future. */
+	Assert_VM_unreachable();
```

Comment:
this will generate core files? should this just be a warning message?

Reply:
It will take a compile change for this line to be run. Ive added an assertion so the developer knows something is wrong.